core/user/user.go

```diff
@@ -67,3 +72,72 @@ func (u UUID) Validate() error {
 func (u UUID) String() string {
 	return string(u)
 }
+
+// userNameTag is the name of the user.
+type userNameTag interface {
+	// Id returns the username string.
+	Id() string
+	// IsLocal indicates if the user is local or external.
+	IsLocal() bool
+}
+
+var (
+	validUserNameSnippet = "[a-zA-Z0-9][a-zA-Z0-9.+-]*[a-zA-Z0-9]"
+	validName            = regexp.MustCompile(fmt.Sprintf("^(?P<name>%s)(?:@(?P<domain>%s))?$", validUserNameSnippet, validUserNameSnippet))
+)
+
+// NewName validates the name and returns a new Name object.
+func NewName(name string) (Name, error) {
+	parts := validName.FindStringSubmatch(name)
+	if len(parts) != 3 {
+		return Name{}, errors.NotValidf("user name %q", name)
+	}
+	domain := parts[2]
+	return Name{
+		name:    name,
+		isLocal: domain == LocalUserDomain || domain == "",
+	}, nil
+}
+
+// Name represents the identity of a user.
+type Name struct {
+	// name is the full username.
+	name string
+	// isLocal indicates if the user is local or external.
+	isLocal bool
+}
+
+// Name returns the full username.
+func (un Name) Name() string {
+	return un.name
+}
+
+// IsLocal indicates if the username is a local or external username.
+func (un Name) IsLocal() bool {
+	return un.isLocal
+}
+
+// String returns the raw username.
+func (un Name) String() string {
+	return un.name
+}
+
+// IsZero return true if the struct is uninitiated.
+func (un Name) IsZero() bool {
+	// The empty string in an invalid user name so the struct is uninitiated if
+	// it is empty.
+	return un.name == ""
+}
+
+// NameFromTag generates a Name from a tag.
+func NameFromTag(tag userNameTag) Name {
+	return Name{
+		name:    tag.Id(),
+		isLocal: tag.IsLocal(),
```

Comment:
names.UserTag also has a domain method, so this will mean the core/user type doesn't lose any information when moving to between the two types if we ever need to do so.

Reply:
We don't actually loose any information because it is all contained within the name string, but its a fair point that we might as well keep the consistant

Reply:
I think that's a good idea.